pageserver/src/tenant/timeline.rs

```diff
@@ -1058,28 +1058,28 @@ impl Timeline {
     pub(crate) async fn wait_lsn(
         &self,
         lsn: Lsn,
-        _ctx: &RequestContext, /* Prepare for use by cancellation */
+        ctx: &RequestContext, /* Prepare for use by cancellation */
     ) -> Result<(), WaitLsnError> {
         if self.cancel.is_cancelled() {
             return Err(WaitLsnError::Shutdown);
         } else if !self.is_active() {
             return Err(WaitLsnError::BadState);
         }
 
-        // This should never be called from the WAL receiver, because that could lead
-        // to a deadlock.
-        debug_assert!(
-            task_mgr::current_task_kind() != Some(TaskKind::WalReceiverManager),
-            "wait_lsn cannot be called in WAL receiver"
-        );
-        debug_assert!(
-            task_mgr::current_task_kind() != Some(TaskKind::WalReceiverConnectionHandler),
-            "wait_lsn cannot be called in WAL receiver"
-        );
-        debug_assert!(
-            task_mgr::current_task_kind() != Some(TaskKind::WalReceiverConnectionPoller),
-            "wait_lsn cannot be called in WAL receiver"
-        );
+        if cfg!(debug_assertions) {
+            match ctx.task_kind() {
+                TaskKind::WalReceiverManager
+                | TaskKind::WalReceiverConnectionHandler
+                | TaskKind::WalReceiverConnectionPoller => {
+                    // This should never be called from the WAL receiver, because that could lead
+                    // to a deadlock.
+                    if let Err(current) = self.last_record_lsn.would_wait_for(lsn) {
+                        panic!("walingest task is calling wait_lsn {lsn} but current is only {current}, would deadlock");
+                    }
+                }
+                _ => {}
+            }
+        }
```

Comment:
The deub-mode test failure of `test_emergency_relocate_with_branches_createdb` is because these assertions are still too conservative;that test exercises the following code path:

https://github.com/neondatabase/neon/blob/97c04855e7c150109761711125de3f343e9dcdfc/pageserver/src/tenant/timeline.rs#L3057-L3064

There, a walreceiver task of the child timelines waits for the parent timeline to catch up. That's not a deadlock risk because the parent timeline walreceiver task can make progress independent of the child timeline.

Wondering if we should just get rid of these assertions.

Reply:
Pushed c0d5cea8be0f5270d1b4544f8b17e8bdd4826de8 , that keeps the assertions at the cost of more ugliness at the call site.  Don't know if it's worth it.